orca/producer_test.go

```diff
@@ -0,0 +1,538 @@
+		Utilization:    map[string]float64{"bob": 555},
+	}
+
+	// Unblock the fake
+	awaitRequest(reportInterval)
+	fake.respCh <- loadReportWant
+	select {
+	case r := <-oobLis.loadReportCh:
+		t.Log("Load report received: ", r)
+		if proto.Equal(r, loadReportWant) {
+			// Success!
+			break
+		}
+	case <-ctx.Done():
+		t.Fatalf("timed out waiting for load report of: %v", loadReportWant)
+	}
+
+	// The next request should be immediate, since there was a message received.
+	expectedBackoff = 99 // This will fail if backoff is called with any other value.
+	fake.respCh <- status.Errorf(codes.Internal, "injected error")
+	awaitRequest(reportInterval)
+
+	// The next requests will need to backoff.
+	expectedBackoff = 0
+	fake.respCh <- status.Errorf(codes.Internal, "injected error")
+	awaitRequest(reportInterval)
+	expectedBackoff = 1
+	fake.respCh <- status.Errorf(codes.Internal, "injected error")
+	awaitRequest(reportInterval)
+	expectedBackoff = 2
+	fake.respCh <- status.Errorf(codes.Internal, "injected error")
+	awaitRequest(reportInterval)
+	expectedBackoff = 99 // This will fail if backoff is called with any other value.
+
+	// Send another valid response and wait for it on the client.
+	fake.respCh <- loadReportWant
+	select {
+	case r := <-oobLis.loadReportCh:
+		t.Log("Load report received: ", r)
+		if proto.Equal(r, loadReportWant) {
+			// Success!
+			break
+		}
+	case <-ctx.Done():
+		t.Fatalf("timed out waiting for load report of: %v", loadReportWant)
+	}
+}
+
+// TestProducerMultipleListeners tests that multiple listeners works as
+// expected in a producer: requesting the proper interval and delivering the
+// update to all listeners.
+func (s) TestProducerMultipleListeners(t *testing.T) {
+	ctx, cancel := context.WithTimeout(context.Background(), defaultTestTimeout)
+	defer cancel()
+
+	// Provide a convenient way to expect backoff calls and return a minimal
+	// value.
+	oldBackoff := internal.DefaultBackoffFunc
+	internal.DefaultBackoffFunc = func(got int) time.Duration {
+		return time.Millisecond
+	}
+	defer func() { internal.DefaultBackoffFunc = oldBackoff }()
+
+	// Initialize listener for our ORCA server.
+	lis, err := testutils.LocalTCPListener()
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	// Register our fake ORCA service
+	s := grpc.NewServer()
+	fake := newFakeORCAService()
+	defer fake.close()
+	v3orcaservicegrpc.RegisterOpenRcaServiceServer(s, fake)
+	go s.Serve(lis)
+	defer s.Stop()
+
+	// Define the report interval and a function to wait for it to be sent to
+	// the server.
+	const reportInterval1 = 123 * time.Second
+	const reportInterval2 = 234 * time.Second
+	const reportInterval3 = 56 * time.Second
+	awaitRequest := func(interval time.Duration) {
+		select {
+		case req := <-fake.reqCh:
+			if got := req.GetReportInterval().AsDuration(); got != interval {
+				t.Errorf("Unexpected report interval; got %v want %v", got, interval)
+			}
+		case <-ctx.Done():
+			panic("Did not receive client request")
```

Comment:
Is it common practice to fail non testing goroutines with a panic rather than log an error?

Reply:
idk if non testing goroutines is the correct language but you get my drift.

Reply:
Yeah was wondering because it was in main test goroutine.